ss/pebbledb/db.go

```diff
@@ -186,13 +187,90 @@ func (db *Database) ApplyChangeset(version int64, cs *proto.NamedChangeSet) erro
 	return b.Write()
 }
 
-// Prune for the PebbleDB SS backend is currently not supported. It seems the only
-// reliable way to prune is to iterate over the desired domain and either manually
-// tombstone or delete. Either way, the operation would be timely.
-//
-// See: https://github.com/cockroachdb/cockroach/blob/33623e3ee420174a4fd3226d1284b03f0e3caaac/pkg/storage/mvcc.go#L3182
+// Prune attempts to prune all versions up to and including the current version
+// Get the range of keys, manually iterate over them and delete them
 func (db *Database) Prune(version int64) error {
-	panic("not implemented!")
+	itr, err := db.storage.NewIter(nil)
+	if err != nil {
+		return err
+	}
+	defer itr.Close()
+
+	batch := db.storage.NewBatch()
+	var counter int
+	var prevKey, prevVersion []byte
+	var prevOK bool
+
+	for itr.First(); itr.Valid(); {
+		// Store current key and version
+		currKey, currVersion, currOK := SplitMVCCKey(itr.Key())
+		if !currOK {
+			return fmt.Errorf("invalid MVCC key")
+		}
+
+		currVer, err := decodeUint64Ascending(currVersion)
+		if err != nil {
+			return err
+		}
+
+		// Seek to next key if we are at a version which is higher than prune height
+		if currVer > version {
+			itr.NextPrefix()
+			continue
+		}
+
+		// Deletes a key if another entry for that key exists a larger version than original but leq prune height
+		if prevOK && bytes.Equal(prevKey, currKey) {
+			prevVer, err := decodeUint64Ascending(prevVersion)
```

Comment:
Why do we need to do the decode again for previous version? Can't we just simply keep the decoded previous version to save some compute time?

Reply:
Updated on latest commit